src/main/scala/com/fulcrumgenomics/bam/EstimatePoolingFractions.scala

```diff
@@ -0,0 +1,242 @@
+    if (this.intervals.isEmpty) {
+      None
+    }
+    else {
+      val list = IntervalList.fromFile(this.intervals.head.toFile)
+      val dict = list.getHeader.getSequenceDictionary
+      this.intervals.tail.foreach { f =>
+        val tmp = IntervalList.fromFile(f.toFile)
+        if (!SequenceUtil.areSequenceDictionariesEqual(dict, tmp.getHeader.getSequenceDictionary))
+          fail(s"Sequence dictionaries differ between ${this.intervals.head} and ${f}")
+        else
+          list.addall(tmp.getIntervals)
+      }
+
+      Some(list.uniqued(false))
+    }
+  }
+
+  /** Generates an iterator over non-filtered bi-allelic SNPs where all the required samples are genotyped. */
+  def constructVcfIterator(in: VCFFileReader, intervals: Option[IntervalList], samples: Array[String]): Iterator[VariantContext] = {
+    val vcfIterator = intervals match {
+      case None     => in.toIterator
+      case Some(is) => ByIntervalListVariantContextIterator(in, is)
+    }
+
+    val samplesAsUtilSet = CollectionUtil.makeSet(samples:_*)
+
+    vcfIterator
+      .filterNot(_.isFiltered)
+      .map(_.subContextFromSamples(samplesAsUtilSet, true))
+      .filter(v => v.isSNP && v.isBiallelic && !v.isMonomorphicInSamples)
+      .filter(_.getNoCallCount == 0)
+      .filter(v => v.getGenotypesOrderedByName.forall(gt => gt.getGQ >= this.minGenotypeQuality))
+  }
+
+  /** Constructs a SamLocusIterator that will visit every locus in the input. */
+  def constructBamIterator(loci: Traversable[Locus]): Iterator[LocusInfo] = {
+    val in = SamReaderFactory.make().open(this.bam)
+    val intervals = new IntervalList(in.getFileHeader)
+    loci.foreach(l => intervals.add(new Interval(l.chrom, l.pos, l.pos)))
+    val iterator = new SamLocusIterator(in, intervals.uniqued())
+    iterator.setEmitUncoveredLoci(true)
+    iterator.setIncludeNonPfReads(false)
+    iterator.setMappingQualityScoreCutoff(this.minMappingQuality)
+    iterator.setQualityScoreCutoff(this.minBaseQuality)
+    iterator.toIterator
+  }
+
+  /**
+    * Fills in the observedFraction field for each locus that meets coverage and then returns
+    * the subset of loci that met coverage.
+    */
+  def fillObserveredFractionAndFilter(loci: Array[Locus], minCoverage: Int): Array[Locus] = {
+    val locusIterator = constructBamIterator(loci)
+    locusIterator.zip(loci.iterator).foreach { case (locusInfo, locus) =>
+      if (locusInfo.getSequenceName != locus.chrom || locusInfo.getPosition != locus.pos) fail("VCF and BAM iterators out of sync.")
+      if (locusInfo.getRecordAndOffsets.size() > minCoverage) {
+        val counts = BaseCounts(locusInfo)
+        val (ref, alt) = (counts(locus.ref), counts(locus.alt))
+
+        // Somewhat redundant with check above, but this protects against a large fraction of Ns or other alleles
+        if (ref + alt >= minCoverage) {
+          locus.observedFraction = Some(alt / (ref + alt).toDouble)
+        }
+      }
+    }
```

Comment:
Why not just use a `flatMap` instead of the `foreach` and subsequent filtering below?